modin/experimental/batch/pipeline.py

```diff
@@ -0,0 +1,386 @@
+
+        Notes
+        -----
+        Use `pandas` for any module level functions inside `func` since it operates directly on
+        partitions.
+        """
+        self.nodes_list.append(
+            PandasQuery(
+                func,
+                is_output,
+                repartition_after,
+                fan_out,
+                pass_partition_id,
+                reduce_fn,
+            )
+        )
+        if is_output:
+            self.outputs.append(self.nodes_list[-1])
+            if output_id:
+                if self.node_to_id is None:
+                    if len(self.outputs) == 1:
+                        self.node_to_id = {}
+                    else:
+                        raise ValueError("Output ID must be specified for all nodes.")
+                self.node_to_id[self.outputs[-1]] = output_id
+            if output_id is None and self.node_to_id is not None:
+                raise ValueError("Output ID must be specified for all nodes.")
+            curr_node = self.outputs[-1]
+            curr_node.operators = self.nodes_list[:-1]
+            self.nodes_list = []
+
+    def _complete_nodes(self, list_of_nodes, partitions):
+        """
+        Run a sub-query end to end.
+
+        Parameters
+        ----------
+        list_of_nodes: list of PandasQuery
+            The functions that compose this query.
+        partitions: list of PandasOnRayDataframeVirtualPartition
+            The partitions that compose the dataframe that is input to this sub-query.
+        """
+        for node in list_of_nodes:
+            if node.fan_out:
+                if len(partitions) > 1:
+                    ErrorMessage.not_implemented(
+                        "Fan out is only supported with DataFrames with 1 partition."
+                    )
+                partitions[0] = partitions[0].force_materialization()
+                ptn_list = partitions[0].list_of_partitions_to_combine
+                partitions[0] = partitions[0].add_to_apply_calls(node.func, 0)
+                partitions[0].drain_call_queue(num_splits=1)
+                oids = []
+                new_dfs = []
+                for i in range(1, self.num_partitions):
+                    new_dfs.append(
+                        type(partitions[0])(
+                            ptn_list,
+                            full_axis=partitions[0].full_axis,
+                        ).add_to_apply_calls(node.func, i)
+                    )
+                    new_dfs[-1].drain_call_queue(num_splits=1)
+                    oids.extend(new_dfs[-1].list_of_blocks)
+                ray.wait(oids, num_returns=len(oids))
+
+                def reducer(df):
+                    df_inputs = [df]
+                    for df in new_dfs:
+                        df_inputs.append(df.to_pandas())
+                    return node.reduce_fn(df_inputs)
+
+                partitions = [partitions[0].add_to_apply_calls(reducer)]
```

Comment:
```suggestion
                other_outputs = [df.to_pandas() for df in new_dfs]
                partitions = [partitions[0].add_to_apply_calls(lambda df: node.reduce_fn([df] + other_outputs))]
```

I think this should work.

Reply:
This blocks in the main thread though - are we sure we want to do this?

Reply:
We shouldn't be blocked in the main thread.